zipline/data/data_portal.py

```diff
@@ -0,0 +1,1186 @@
+
+log = Logger('DataPortal')
+
+HISTORY_FREQUENCIES = ["1d", "1m"]
+
+BASE_FIELDS = {
+    'open': 'open',
+    'open_price': 'open',
+    'high': 'high',
+    'low': 'low',
+    'close': 'close',
+    'close_price': 'close',
+    'volume': 'volume',
+    'price': 'close'
+}
+
+
+class DataPortal(object):
+    def __init__(self,
+                 env,
+                 sim_params=None,
+                 minutes_equities_path=None,
+                 daily_equities_path=None,
+                 adjustment_reader=None,
+                 sid_path_func=None):
+        self.env = env
+
+        # Internal pointers to the current dt (can be minute) and current day.
+        # In daily mode, they point to the same thing. In minute mode, it's
+        # useful to have separate pointers to the current day and to the
+        # current minute.  These pointers are updated by the
+        # AlgorithmSimulator's transform loop.
+        self.current_dt = None
+        self.current_day = None
+
+        # This is a bit ugly, but is here for performance reasons.  In minute
+        # simulations, we need to very quickly go from dt -> (# of minutes
+        # since Jan 1 2002 9:30 Eastern).
+        #
+        # The clock that heartbeats the simulation has all the necessary
+        # information to do this calculation very quickly.  This value is
+        # calculated there, and then set here
+        self.cur_data_offset = 0
+
+        self.views = {}
+
+        if minutes_equities_path is None and daily_equities_path is None:
+            raise ValueError("Must provide at least one of minute or "
+                             "daily data path!")
+
+        self._minutes_equities_path = minutes_equities_path
+        self._daily_equities_path = daily_equities_path
+        self._asset_finder = env.asset_finder
+
+        self._carrays = {
+            'open': {},
+            'high': {},
+            'low': {},
+            'close': {},
+            'volume': {},
+            'sid': {},
+            'dt': {},
+        }
+
+        self._adjustment_reader = adjustment_reader
+
+        # caches of sid -> adjustment list
+        self._splits_dict = {}
+        self._mergers_dict = {}
+        self._dividends_dict = {}
+
+        # Pointer to the daily bcolz file.
+        self._daily_equities_data = None
+
+        # Cache of sid -> the first trading day of an asset, even if that day
+        # is before 1/2/2002.
+        self._asset_start_dates = {}
+        self._asset_end_dates = {}
+
+        # Fetcher state
+        self._augmented_sources_map = {}
+        self._fetcher_df = None
+
+        self._sim_params = sim_params
+        if self._sim_params is not None:
+            self._data_frequency = self._sim_params.data_frequency
+
+        self._sid_path_func = sid_path_func
+
+        self.DAILY_PRICE_ADJUSTMENT_FACTOR = 0.001
+        self.MINUTE_PRICE_ADJUSTMENT_FACTOR = 0.001
+
+    def handle_extra_source(self, source_df):
```

Comment:
I think calling this multiple times will cause `self._fetcher_df` to get reset, but other state, like the `augmented_sources_map` will accrue more data.  Probably we don't want them out of sync?